astropy/modeling/tests/test_quantities_parameters.py

```diff
@@ -0,0 +1,476 @@
+                                 "a Quantity with units equivalent to Jy")
+
+    with pytest.raises(UnitsError) as exc:
+        g.amplitude = 2 * u.s
+    assert exc.value.args[0] == ("The 'amplitude' parameter should be given as "
+                                 "a Quantity with units equivalent to Jy")
+
+    with pytest.raises(UnitsError) as exc:
+        g.mean = 2 * u.Jy
+    assert exc.value.args[0] == ("The 'mean' parameter should be given as a "
+                                 "unitless value")
+
+
+def test_parameter_set_incompatible_units_with_equivalencies():
+    """
+    Check that parameters can only be set to quantities with equivalent units,
+    taking into account equivalencies defined in the parameter initializers.
+    """
+
+    class TestModel(BaseTestModel):
+        a = Parameter(default=1.0, unit=u.m)
+        b = Parameter(default=1.0, unit=u.m, equivalencies=u.spectral())
+        c = Parameter(default=1.0, unit=u.K, equivalencies=u.temperature_energy())
+
+    model = TestModel()
+
+    with pytest.raises(UnitsError) as exc:
+        model.a = 3 * u.Hz
+    assert exc.value.args[0] == "The 'a' parameter should be given as a Quantity with units equivalent to m"
+
+    model.b = 3 * u.Hz
+
+    with pytest.raises(UnitsError) as exc:
+        model.c = 3 * u.Hz
+    assert exc.value.args[0] == "The 'c' parameter should be given as a Quantity with units equivalent to K"
+
+    model.c = 3 * u.eV
+
+
+def test_parameter_equivalencies_used_in_init():
+    """
+    Check that equivalencies are taken into account when initializing a model
+    with quantities, if using a model where the parameters have specific
+    equivalencies defined.
+    """
+
+    class TestModel(Model):
+        a = Parameter(default=1.0, unit=u.m)
+        b = Parameter(default=1.0, unit=u.m, equivalencies=u.spectral())
+        @staticmethod
+        def evaluate(x, a):
+            return x
+
+    with pytest.raises(InputParameterError) as exc:
+        model = TestModel(a=2 * u.Hz)
+    assert exc.value.args[0] == ('TestModel.__init__() requires parameter \'a\' to '
+                                 'be in units equivalent to Unit("m") (got Unit("Hz"))')
+
+    model = TestModel(b=2 * u.Hz)
+
+    with pytest.raises(InputParameterError) as exc:
+        TestModel(b=2 * u.s)
+    assert exc.value.args[0] == ('TestModel.__init__() requires parameter \'b\' to '
+                                 'be in units equivalent to Unit("m") (got Unit("s"))')
+
+
+def test_parameter_set_equivalency():
+    """
+    Test that we can set equivalencies on an existing model.
+    """
+    g = Gaussian1D(1 * u.Jy, 3 * u.m, 0.1 * u.nm)
+
+    with pytest.raises(UnitsError) as exc:
+        g.mean = 3 * u.Hz
+    assert exc.value.args[0] == ("The 'mean' parameter should be given as a "
+                                 "Quantity with units equivalent to m")
+
+    g.mean.equivalencies = u.spectral()
```

Comment:
Since equivalency functions return lists, I think you can just add them together? But we could consider supporting lists of lists too.